libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1819,19 +1821,60 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 				if gt(length, 0xffffffffffffffff) { <panic>() }
 				<resizeArray>(dst, length)
 
-				let srcPtr := <srcDataLocation>(src)
-
-				let dstPtr := <dstDataLocation>(dst)
+				let srcSlot := <srcDataLocation>(src)
+				let dstSlot := <dstDataLocation>(dst)
 
 				let fullSlots := div(length, <itemsPerSlot>)
-				let i := 0
-				for { } lt(i, fullSlots) { i := add(i, 1) } {
-					sstore(add(dstPtr, i), <maskFull>(sload(add(srcPtr, i))))
-				}
-				let spill := sub(length, mul(i, <itemsPerSlot>))
-				if gt(spill, 0) {
-					sstore(add(dstPtr, i), <maskBytes>(sload(add(srcPtr, i)), mul(spill, <bytesPerItem>)))
+
+				let srcSlotValue := sload(srcSlot)
+				let srcItemIndexInSlot := 0
+				for { let i := 0 } lt(i, fullSlots) { i := add(i, 1) } {
+					let dstSlotValue := 0
+					<?sameType>
+						dstSlotValue := <maskFull>(srcSlotValue)
+						<updateSrcSlotValue>
+					<!sameType>
+						<?multipleItemsPerSlotDst>for { let j := 0 } lt(j, <itemsPerSlot>) { j := add(j, 1) } </multipleItemsPerSlotDst>
```

Comment:
In case that we do not have multiple items per slot in destination, there is no loop. Maybe representation is ugly, but the body of the loop is almost the same so didn't want to put whole loop under condition...